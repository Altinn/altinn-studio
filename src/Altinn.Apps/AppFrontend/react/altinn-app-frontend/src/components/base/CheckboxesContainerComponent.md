Simple checkbox with 3 options

```tsx
import {CheckboxContainerComponent} from './CheckboxesContainerComponent';
import Legend from '../../features/form/components/Legend';
import {nb} from '../../shared/resources/language/texts/nb';
const dummyFunc = () => {return;}

const legend = () => {
  return (
    <Legend
      labelText={'Simple checkboxes'}
      language={nb()}
      required={false}
      helpTextProps={{}}
    />
  );
}

const props = {
  id: 'simpleCheckbox',
  formData: '',
  handleDataChange: dummyFunc,
  handleFocusUpdate: dummyFunc,
  isValid: true,
  validationMessages: {},
  options: [
    {
      label: 'Label 1',
      value: 'v1',
    },
    {
      label: 'Label 2',
      value: 'v2',
    },
    {
      label: 'Label 3',
      value: 'v3',
    },
  ],
  preselectedOptionIndex: null,
  readOnly: false,
  shouldFocus: false,
  legend,
};

<CheckboxContainerComponent
  {...props}
/>

```

