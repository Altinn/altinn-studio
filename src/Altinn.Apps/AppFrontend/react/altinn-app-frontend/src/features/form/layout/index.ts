import { IOption, Triggers } from '../../../types';

export interface ILayouts {
  [id: string]: ILayout;
}

export interface ILayoutEntry {
  id: string;
  type?: string;
}

export interface ILayoutGroup extends ILayoutEntry {
  children: string[];
  dataModelBindings?: IDataModelBindings;
  maxCount: number;
  textResourceBindings?: ITextResourceBindings;
  tableHeaders?: string[];
}

export interface ILayoutComponent extends ILayoutEntry {
  type: string;
  dataModelBindings: IDataModelBindings;
  isValid?: boolean;
  readOnly: boolean;
  disabled?: boolean;
  required: boolean;
  textResourceBindings: ITextResourceBindings;
  triggers?: Triggers[];
}
export interface IDataModelBindings {
  [id: string]: string;
}

export interface ITextResourceBindings {
  [id: string]: string;
}

export type ILayout = Array<ILayoutComponent | ILayoutGroup>;

export interface ISelectionComponentProps extends ILayoutComponent {
  options?: IOption[];
  optionsId?: string;
}

export interface IDatepickerProps extends ILayoutComponent { }

export interface IFileuploadProps extends ILayoutComponent {
  maxNumberOfAttachments: number;
  maxFileSizeInMB: number;
  displayMode: any;
  hasCustomFileEndings: boolean;
  validFileEndings: any;
}

declare enum HeaderSize {
  S,
  M,
  L,
}

export interface IHeaderProps extends ILayoutComponent {
  size: HeaderSize;
}

declare enum InputFieldType {
  text,
  email,
  password,
}

export interface IInputProps extends ILayoutComponent {
  inputType: InputFieldType;
}

export interface INavigationButtonProps extends ILayoutComponent {
  next?: string;
  previous?: string;
}

export interface IParagraphProps extends ILayoutComponent { }

export interface IRadioButtonsProps extends ILayoutComponent {
  options: IOption[];
  preselectedOptionIndex: number;
}

export interface ITextAreaProps extends ILayoutComponent { }

export interface IAdressComponent extends ILayoutComponent {
  addressTextResourceBinding: string;
  areaCodeTextResourceBinding: string;
  coTextResourceBinding: string;
  simpleDisplayMode: boolean;
}
