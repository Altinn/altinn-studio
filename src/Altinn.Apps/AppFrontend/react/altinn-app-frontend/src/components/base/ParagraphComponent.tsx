import * as React from 'react';
import { Grid, makeStyles, Typography } from '@material-ui/core';
import { ITextResourceBindings } from 'src/features/form/layout';
import { HelpTextContainer } from 'src/features/form/components/HelpTextContainer';

export interface IParagraphProps {
  id: string;
  text: string;
  textResourceBindings: ITextResourceBindings;
  language: any;
  getTextResource: (key: string) => string;
}

const useStyles = makeStyles({
  spacing: {
    letterSpacing: '0.3px',
  },
  // Class to override default stylings for headers created by markdown parsing. Done to align help text icon.
  typography: {
    '& h1': {
      margin: 0,
    },
    '& h2': {
      margin: 0,
    },
    '& h3': {
      margin: 0,
    },
    '& h4': {
      margin: 0,
    },
    '& h5': {
      margin: 0,
    },
    '& h6': {
      margin: 0,
    },
  },
});

export function ParagraphComponent(props: IParagraphProps) {
  const classes = useStyles();
  return (
    <Grid
      container={true}
      direction='row'
      alignItems='center'
    >
      <Grid item={true}>
        <Typography
          id={props.id}
          className={`${classes.spacing} ${classes.typography}`}
        >
          {props.text}
        </Typography>
      </Grid>
      {props.textResourceBindings?.help &&
      <Grid item={true} className={classes.spacing}>
        <HelpTextContainer
          language={props.language}
          id={props.id}
          helpText={props.getTextResource(props.textResourceBindings.help)}
        />
      </Grid>
      }
    </Grid>
  );
}
