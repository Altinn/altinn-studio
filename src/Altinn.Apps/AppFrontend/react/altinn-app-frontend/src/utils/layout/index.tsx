/* eslint-disable max-len */
import * as React from 'react';
import { GroupContainer } from 'src/features/form/containers/GroupContainer';
// eslint-disable-next-line import/no-cycle
import { IInstance } from 'altinn-shared/types';
import { GenericComponent } from '../../components/GenericComponent';
import { ILayouts, ILayoutComponent, ILayoutGroup, ILayout } from '../../features/form/layout';
import { ILayoutSets, ILayoutSet } from '../../types';

export function getLayoutComponentById(id: string, layouts: ILayouts): ILayoutComponent {
  let component: ILayoutComponent;
  Object.keys(layouts).forEach((layoutId) => {
    if (!component) {
      component = layouts[layoutId].find((element) => {
        // Check against provided id, with potential -{index} postfix.
        const match = matchLayoutComponent(id, element.id);
        return match && match.length > 0;
      }) as ILayoutComponent;
    }
  });

  return component;
}

export function getLayoutIdForComponent(id: string, layouts: ILayouts): string {
  let foundLayout: string;
  Object.keys(layouts).forEach((layoutId) => {
    if (!foundLayout) {
      const component = layouts[layoutId].find((element) => {
        // Check against provided id, with potential -{index} postfix.
        const match = matchLayoutComponent(id, element.id);
        return match && match.length > 0;
      }) as ILayoutComponent;
      if (component) {
        foundLayout = layoutId;
      }
    }
  });
  return foundLayout;
}

/*
  Check if provided id matches component id.
  For repeating groups, component id from formLayout is postfixed with -{index}
  when rendering, where index is the component's index (number) in the repeating group list.
  This does not change the component definition in formLayout.
  Therefore, we must match on component id as well as a potential -{index} postfix
  when searching through formLayout for the component definition.
*/
export function matchLayoutComponent(providedId: string, componentId: string) {
  return providedId.match(`${componentId}(-[0-9]*)*$`);
}

export function renderGenericComponent(component: ILayoutComponent, layout: ILayout, index: number = -1) {
  if (component.type.toLowerCase() === 'group') {
    return renderLayoutGroup(component as unknown as ILayoutGroup, layout, index);
  }
  return (
    <GenericComponent
      key={component.id}
      {...component}
    />
  );
}

export function renderLayoutGroup(layoutGroup: ILayoutGroup, layout: ILayout, index?: number) {
  const groupComponents = layoutGroup.children.map((child) => {
    return layout.find((c) => c.id === child) as ILayoutComponent;
  });
  const deepCopyComponents = setupGroupComponents(groupComponents, layoutGroup.dataModelBindings.group, index);
  const repeating = layoutGroup.maxCount > 1;
  if (!repeating) {
    // If not repeating, treat as regular components
    return (
      <>
        {deepCopyComponents.map((component: ILayoutComponent) => { return renderGenericComponent(component, layout); })}
      </>
    );
  }

  return (
    <GroupContainer
      container={layoutGroup}
      id={layoutGroup.id}
      key={layoutGroup.id}
      components={deepCopyComponents}
    />
  );
}

export function setupGroupComponents(components: (ILayoutComponent | ILayoutGroup)[], groupDataModelBinding: string, index: number): (ILayoutGroup | ILayoutComponent)[] {
  const childComponents = components.map((component: ILayoutComponent) => {
    const componentDeepCopy: ILayoutComponent = JSON.parse(JSON.stringify(component));
    const dataModelBindings = { ...componentDeepCopy.dataModelBindings };
    Object.keys(dataModelBindings).forEach((key) => {
      // eslint-disable-next-line no-param-reassign
      const originalGroupBinding = groupDataModelBinding.replace(`[${index}]`, '');
      dataModelBindings[key] = dataModelBindings[key].replace(originalGroupBinding, groupDataModelBinding);
    });
    const deepCopyId = `${componentDeepCopy.id}-${index}`;

    return {
      ...componentDeepCopy,
      dataModelBindings,
      id: deepCopyId,
      baseComponentId: componentDeepCopy.id,
    };
  });
  return childComponents;
}

export function getLayoutsetForDataElement(instance: IInstance, datatype: string, layoutsets: ILayoutSets) {
  const currentTaskId = instance.process.currentTask.elementId;
  const foundLayout = layoutsets.sets.find((layoutSet: ILayoutSet) => {
    if (layoutSet.dataType !== datatype) {
      return false;
    }
    return layoutSet.tasks.find((taskId: string) => taskId === currentTaskId);
  });
  return foundLayout.id;
}

export function getHiddenFieldsForGroup(hiddenFields: string[], components: (ILayoutGroup | ILayoutComponent)[]) {
  const result = [];
  hiddenFields.forEach((fieldKey) => {
    const fieldKeyWithoutIndex = fieldKey.replace(/-\d{1,}$/, '');
    if (components.find((component) => component.id === fieldKeyWithoutIndex)) {
      result.push(fieldKey);
    }
  });

  return result;
}
