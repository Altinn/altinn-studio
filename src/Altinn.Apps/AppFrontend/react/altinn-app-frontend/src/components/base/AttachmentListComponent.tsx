import * as React from 'react';
import { AltinnAttachment } from 'altinn-shared/components';
import { useSelector } from 'react-redux';
import { IRuntimeState } from 'src/types';
import { IAttachment } from 'altinn-shared/types';
import { mapInstanceAttachments } from 'altinn-shared/utils';
import { Grid, Typography } from '@material-ui/core';

export interface IAttachmentListProps {
  id: string;
  text: string;
  dataTypeIds?: string[];
}

export function AttachmentListComponent(props: IAttachmentListProps) {
  const currentTaskId = useSelector(
    (state: IRuntimeState) => state.instanceData.instance.process.currentTask.elementId,
  );
  const dataForTask = useSelector((state: IRuntimeState) => {
    const dataTypes = state.applicationMetadata.applicationMetadata.dataTypes.filter((type) => {
      return type.taskId === state.instanceData.instance.process.currentTask.elementId;
    });
    return state.instanceData.instance.data.filter((dataElement) => {
      if (props.dataTypeIds) {
        return props.dataTypeIds.findIndex((id) => dataElement.dataType === id) > -1;
      }
      return dataTypes.findIndex((type) => dataElement.dataType === type.id) > -1;
    });
  });
  const attachments: IAttachment[] = useSelector((state: IRuntimeState) => {
    const appLogicDataTypes = state.applicationMetadata.applicationMetadata.dataTypes.filter((dataType) => {
      return dataType.appLogic && dataType.taskId === currentTaskId;
    });
    return mapInstanceAttachments(dataForTask, appLogicDataTypes.map((type) => type.id));
  });

  return (
    <Grid item={true} xs={12}>
      <Typography className='attachmentList-title' component='span'>
        {props.text || ''}
      </Typography>
      <AltinnAttachment
        attachments={attachments}
      />
    </Grid>
  );
}
