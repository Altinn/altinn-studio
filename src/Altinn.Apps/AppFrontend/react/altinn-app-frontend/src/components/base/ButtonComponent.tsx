/* eslint-disable react/prop-types */
import * as React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getLanguageFromKey } from 'altinn-shared/utils/language';
import { AltinnLoader } from 'altinn-shared/components';
import { IAltinnWindow, IRuntimeState } from '../../types';
import FormDataActions from '../../features/form/data/formDataActions';

export interface IButtonProvidedProps {
  id: string;
  text: string;
  disabled: boolean;
  handleDataChange: (value: any) => void;
  formDataCount: number;
  language: any;
}

export function ButtonComponent(props: IButtonProvidedProps) {
  const dispatch = useDispatch();
  const autoSave = useSelector((state: IRuntimeState) => state.formLayout.uiConfig.autoSave);
  const isSubmitting = useSelector((state: IRuntimeState) => state.formData.isSubmitting);
  const isSaving = useSelector((state: IRuntimeState) => state.formData.isSaving);
  const ignoreWarnings = useSelector((state: IRuntimeState) => state.formData.ignoreWarnings);

  const renderSubmitButton = () => {
    return (
      <div className='col-4 pl-0'>
        {isSubmitting ?
          renderLoader() :
          <button
            type='submit'
            className='a-btn a-btn-success'
            onClick={submitForm}
            id={props.id}
            style={{ marginBottom: '0' }}
          >
            {props.text}
          </button>
        }
      </div>
    );
  };

  const renderSaveButton = () => {
    return (
      <div className='col-2 pl-0'>
        {isSaving ?
          renderLoader() :
          <button
            type='submit'
            className='a-btn a-btn-success'
            onClick={saveFormData}
            id='saveBtn'
            style={{ marginBottom: '0' }}
          >
            Lagre
          </button>
        }
      </div>
    );
  };

  const saveFormData = () => {
    dispatch(FormDataActions.submitFormData({}));
  };

  const renderLoader = () => {
    return (
      <AltinnLoader
        srContent={getLanguageFromKey('general.loading', props.language)}
        style={{
          marginLeft: '40px',
          marginTop: '2px',
          height: '45px', // same height as button
        }}
      />
    );
  };

  const submitForm = () => {
    const {
      org, app, instanceId,
    } = window as Window as IAltinnWindow;
    dispatch(FormDataActions.submitFormData({
      url: `${window.location.origin}/${org}/${app}/api/${instanceId}`,
      apiMode: 'Complete',
      stopWithWarnings: !ignoreWarnings,
    }));
  };

  return (
    <div className='container pl-0'>
      <div className='a-btn-group' style={{ marginTop: '3.6rem', marginBottom: '0' }}>
        <div className='row' style={{ marginLeft: '0' }}>
          {autoSave === false && renderSaveButton()}
          {renderSubmitButton()}
        </div>
      </div>
    </div>
  );
}
