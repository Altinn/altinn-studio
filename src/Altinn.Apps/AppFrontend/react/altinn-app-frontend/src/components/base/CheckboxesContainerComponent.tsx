/* eslint-disable import/first */
/* eslint-disable react/no-array-index-key */
/* eslint-disable no-undef */
import { FormControlLabel, FormGroup, FormLabel } from '@material-ui/core';
import Checkbox, { CheckboxProps } from '@material-ui/core/Checkbox';
import FormControl from '@material-ui/core/FormControl';
import { makeStyles } from '@material-ui/core/styles';
import * as React from 'react';
import { AltinnAppTheme } from 'altinn-shared/theme';
import classNames from 'classnames';
import { useSelector } from 'react-redux';
import { IRuntimeState } from 'src/types';
import { renderValidationMessagesForComponent } from '../../utils/render';

export interface ICheckboxContainerProps {
  id: string;
  formData: any;
  handleDataChange: (value: any) => void;
  handleFocusUpdate: (value: any) => void;
  isValid: boolean;
  validationMessages: any;
  options: any[];
  optionsId: string;
  preselectedOptionIndex: number;
  readOnly: boolean;
  shouldFocus: boolean;
  legend: () => JSX.Element;
  getTextResource: (key: string) => JSX.Element;
  getTextResourceAsString: (key: string) => string;
}

export interface IStyledCheckboxProps extends CheckboxProps {
  label: string;
}

const useStyles = makeStyles({
  root: {
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
  icon: {
    border: '2px solid #1EAEF7',
    width: 24,
    height: 24,
    backgroundColor: '#ffffff',
    '$root.Mui-focusVisible &': {
      outline: '2px solid #ff0000',
      outlineOffset: 0,
      outlineColor: AltinnAppTheme.altinnPalette.primary.blueDark,
    },
    'input:hover ~ &': {
      borderColor: AltinnAppTheme.altinnPalette.primary.blueDark,
    },
    'input:disabled ~ &': {
      boxShadow: 'none',
      background: 'rgba(206,217,224,.5)',
    },
  },
  checkedIcon: {
    backgroundColor: '#ffffff',
    '&:before': {
      display: 'block',
      width: 20,
      height: 20,
      // tslint:disable-next-line: max-line-length
      backgroundImage: "url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'%3E%3Cpath fill='%23000000' d='M6.564.75l-3.59 3.612-1.538-1.55L0 4.26 2.974 7.25 8 2.193z'/%3E%3C/svg%3E\")",
      content: '""',
    },
    'input:hover ~ &': {
      borderColor: AltinnAppTheme.altinnPalette.primary.blueDark,
    },
  },
  legend: {
    color: '#000000',
  },
  margin: {
    marginBottom: '1.2rem',
  },
});

function usePrevious(value) {
  const ref = React.useRef();
  React.useEffect(() => {
    ref.current = value.slice();
  });

  return ref.current;
}

export const CheckboxContainerComponent = (props: ICheckboxContainerProps) => {
  const classes = useStyles(props);
  const apiOptions = useSelector((state: IRuntimeState) => state.optionState.options[props.optionsId]);
  const options = apiOptions || props.options || [];
  const [selected, setSelected] = React.useState([]);
  const prevSelected: any = usePrevious(selected);
  const checkBoxesIsRow: boolean = (options.length <= 2);

  React.useEffect(() => {
    returnState();
  }, [options]);

  React.useEffect(() => {
    returnState();
  }, [props.formData]);

  const returnState = () => {
    if (
      !props.formData &&
      props.preselectedOptionIndex >= 0 &&
      options &&
      props.preselectedOptionIndex < options.length
    ) {
      const preSelected: string[] = [];
      preSelected[props.preselectedOptionIndex] = options[props.preselectedOptionIndex].value;
      props.handleDataChange(preSelected[props.preselectedOptionIndex]);
      setSelected(preSelected);
    } else {
      setSelected(props.formData ? props.formData.toString().split(',') : []);
    }
  };

  const onDataChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newSelected: any = selected.slice();
    const index = newSelected.indexOf(event.target.name);

    if (index >= 0) {
      newSelected[index] = '';
    } else {
      newSelected.push(event.target.name);
    }
    const filtered = newSelected.filter((element: string) => !!element);
    props.handleFocusUpdate(props.id);
    props.handleDataChange(selectedHasValues(filtered) ? filtered.join() : '');
  };

  const handleOnBlur = () => {
    props.handleDataChange(props.formData);
  };

  const selectedHasValues = (select: string[]): boolean => {
    return select.some((element) => element !== '');
  };

  const isOptionSelected = (option: string) => {
    return selected.indexOf(option) > -1;
  };

  const inFocus = (index: number) => {
    let changed: any;
    if (!prevSelected) {
      return false;
    }
    if (prevSelected.length === 0) {
      changed = selected.findIndex((x) => !!x && x !== '');
    } else {
      changed = selected.findIndex((x) => !prevSelected.includes(x));
    }
    if (changed === -1) {
      changed = prevSelected.findIndex((x) => !selected.includes(x));
    }

    if (changed === -1) {
      return false;
    }
    const should = props.shouldFocus && changed === index;
    return should;
  };

  const RenderLegend = props.legend;

  return (
    <FormControl key={`checkboxes_control_${props.id}`} component='fieldset'>
      <FormLabel component='legend' classes={{ root: classNames(classes.legend) }}>
        <RenderLegend />
      </FormLabel>
      <FormGroup
        row={checkBoxesIsRow}
        id={props.id}
        key={`checkboxes_group_${props.id}`}
      >
        {options.map((option, index) => (
          <React.Fragment key={option.value}>
            <FormControlLabel
              key={option.value}
              classes={{ root: classNames(classes.margin) }}
              control={(
                <StyledCheckbox
                  checked={isOptionSelected(option.value)}
                  onChange={onDataChanged}
                  onBlur={handleOnBlur}
                  value={index}
                  key={option.value}
                  name={option.value}
                  autoFocus={inFocus(index)}
                  label={props.getTextResourceAsString(option.label)}
                />
              )}
              label={props.getTextResource(option.label)}
            />
            { props.validationMessages &&
              isOptionSelected(option.value) &&
              renderValidationMessagesForComponent(props.validationMessages.simpleBinding, props.id) }
          </React.Fragment>
        ))}
      </FormGroup>
    </FormControl>
  );
};

const StyledCheckbox = (styledCheckboxProps: IStyledCheckboxProps) => {
  const { label, ...checkboxProps } = styledCheckboxProps;
  const classes = useStyles(styledCheckboxProps);

  return (
    <Checkbox
      className={classes.root}
      disableRipple={true}
      color='default'
      checkedIcon={<span className={classNames(classes.icon, classes.checkedIcon)} />}
      icon={<span className={classes.icon} />}
      inputProps={{ 'aria-label': label }}
      {...checkboxProps}
    />
  );
};
