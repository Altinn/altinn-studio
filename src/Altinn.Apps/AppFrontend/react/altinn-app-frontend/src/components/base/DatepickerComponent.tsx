import moment from 'moment';
import * as React from 'react';
import '../../styles/DatepickerComponent.css';
import '../../styles/shared.css';
import { Grid, useMediaQuery, useTheme, Icon, makeStyles } from '@material-ui/core';
import { KeyboardDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';
import { getLanguageFromKey } from 'altinn-shared/utils';
import { AltinnAppTheme } from 'altinn-shared/theme';
import { Moment } from 'moment';
import { IComponentValidations, IComponentBindingValidation } from 'src/types/global';
import { renderValidationMessagesForComponent } from '../../utils/render';
import { validateDatepickerFormData } from '../../utils/validation';

export interface IDatePickerProps{
  id: string;
  readOnly: boolean;
  required: boolean;
  formData: any;
  handleDataChange: (value: any) => void;
  isValid?: boolean;
  format: string;
  minDate: string;
  maxDate: string;
  language: any;
  componentValidations: IComponentValidations;
}

const useStyles = makeStyles({
  root: {
    fontSize: '1.6rem',
    borderWidth: '2px',
    borderStyle: 'solid',
    marginBottom: '0px',
    borderColor: AltinnAppTheme.altinnPalette.primary.blueMedium,
    '&:hover': {
      borderColor: AltinnAppTheme.altinnPalette.primary.blueDark,
    },
    '&:focus-within': {
      outlineOffset: '0px',
      outline: `2px solid ${AltinnAppTheme.altinnPalette.primary.blueDark}`,
    },
  },
  input: {
    marginLeft: '12px',
  },
  invalid: {
    borderColor: `${AltinnAppTheme.altinnPalette.primary.red} !important`,
    outlineColor: `${AltinnAppTheme.altinnPalette.primary.red} !important`,
  },
  icon: {
    // must be the same to center align icon
    fontSize: '30px',
    lineHeight: '30px',
  },
  formHelperText: {
    fontSize: '1.4rem',
  },
  datepicker: {
    width: 'auto',
    marginBottom: '0px',
    marginTop: '0px',
  },
});

class AltinnMomentUtils extends MomentUtils {
  getDatePickerHeaderText(date: Moment) {
    if (date && date.locale() == 'nb') {
      return date.format('ddd, D MMM');
    }
    return super.getDatePickerHeaderText(date);
  }
}

export const DatePickerMinDateDefault = '1900-01-01T12:00:00.000Z';
export const DatePickerMaxDateDefault = '2100-01-01T12:00:00.000Z';
export const DatePickerFormatDefault = 'DD/MM/YYYY';

function DatepickerComponent(props: IDatePickerProps) {
  const classes = useStyles();
  const [date, setDate] = React.useState<moment.Moment>(null);
  const [validDate, setValidDate] = React.useState<boolean>(true);
  const [validationMessages, setValidationMessages] = React.useState<IComponentBindingValidation>(null);
  const minDate = props.minDate || DatePickerMinDateDefault;
  const maxDate = props.maxDate || DatePickerMaxDateDefault;
  const format = props.format || DatePickerFormatDefault;

  const locale = window.navigator?.language || (window.navigator as any)?.userLanguage || 'nb-NO';
  moment.locale(locale);
  const theme = useTheme();
  const inline = useMediaQuery(theme.breakpoints.up('sm'));

  const isDateEmpty = () => {
    return date && date.parsingFlags().parsedDateParts.length === 0;
  };

  const getValidationMessages = () => {
    const checkDate = isDateEmpty() ? '' : date?.toISOString();
    const validations: IComponentBindingValidation =
      validateDatepickerFormData(checkDate, minDate, maxDate, format, props.language);
    const suppliedValidations = props.componentValidations?.simpleBinding;
    if (suppliedValidations?.errors) {
      suppliedValidations.errors.forEach((validation: string) => {
        if (validations.errors.indexOf(validation) == -1) {
          validations.errors.push(validation);
        }
      });
    }
    if (suppliedValidations?.warnings) {
      suppliedValidations.warnings.forEach((validation: string) => {
        if (validations.warnings.indexOf(validation) == -1) {
          validations.warnings.push(validation);
        }
      });
    }
    return validations;
  };

  React.useEffect(() => {
    const date = moment(props.formData || '');
    setDate(date);
  }, [props.formData]);

  React.useEffect(() => {
    setValidationMessages(getValidationMessages());
  }, [props.formData, props.componentValidations]);

  const handleDataChangeWrapper = (date: moment.Moment) => {
    setDate(date);
    setValidDate(true); // we reset valid date => show error onBlur or when user is done typing
    setValidationMessages({});
    if (date && date.isValid()) {
      props.handleDataChange(date?.toISOString());
      setValidDate(isValidDate(date)); // the date can have a valid format but not pass min/max validation
    }
  };

  const isValidDate = (date: moment.Moment): boolean => {
    if (!date) {
      return true;
    }
    return date.isValid() && date.isAfter(minDate) && date.isBefore(maxDate);
  };

  const handleOnBlur = () => {
    setValidDate(isValidDate(date));
    setValidationMessages(getValidationMessages());
    const saveDate = isDateEmpty() ? '' : date?.toISOString();
    props.handleDataChange(saveDate);
  };

  return (
    <>
      <MuiPickersUtilsProvider utils={AltinnMomentUtils}>
        <Grid
          container
          item
          xs={12}
        >
          <KeyboardDatePicker
            readOnly={props.readOnly}
            required={props.required}
            variant={inline ? 'inline' : 'dialog'}
            format={format}
            margin='normal'
            id={props.id}
            value={date}
            placeholder={format}
            key={props.id}
            onChange={handleDataChangeWrapper}
            onBlur={handleOnBlur}
            autoOk={true}
            invalidDateMessage='' // all validation messages intentionally left empty
            maxDateMessage=''
            minDateMessage=''
            minDate={minDate}
            maxDate={maxDate}
            cancelLabel={getLanguageFromKey('date_picker.cancel_label', props.language)}
            clearLabel={getLanguageFromKey('date_picker.clear_label', props.language)}
            todayLabel={getLanguageFromKey('date_picker.today_label', props.language)}
            InputProps={{
              disableUnderline: true,
              'aria-describedby': `description-${props.id}`,
              error: (!props.isValid || !validDate),
              classes: {
                root: classes.root + ((!props.isValid || !validDate) ? ` ${classes.invalid}` : ''),
                input: classes.input,
              },
            }}
            FormHelperTextProps={{
              classes: {
                root: classes.formHelperText,
              },
            }}
            KeyboardButtonProps={{
              'aria-label': getLanguageFromKey('date_picker.aria_label_icon', props.language),
            }}
            keyboardIcon={
              <Icon
                className={`${classes.icon} ai ai-date`}
              />
            }
            className={classes.datepicker}
          />
        </Grid>
      </MuiPickersUtilsProvider>
      {renderValidationMessagesForComponent(validationMessages, `${props.id}_validations`)}
    </>
  );
}

export default DatepickerComponent;
