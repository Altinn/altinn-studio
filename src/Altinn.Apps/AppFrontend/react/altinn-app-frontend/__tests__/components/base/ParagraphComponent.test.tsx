/* tslint:disable:jsx-wrap-multiline */
import { shallow } from 'enzyme';
import 'jest';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ParagraphComponent } from '../../../src/components/base/ParagraphComponent';
import { ITextResourceBindings } from '../../../src/features/form/layout';

describe('>>> components/base/ParagraphComponent.tsx --- Snapshot', () => {
  let mockId: string;
  let mockText: string;
  let mockGetTextResource: (key: string) => string;
  let mockLanguage: any;
  let mockTextResourceBindings: ITextResourceBindings;
  mockId = 'mock-id';
  mockText = 'Here goes a paragraph';
  mockGetTextResource = (key: string) => key;
  mockLanguage = {};
  mockTextResourceBindings = {
    tile: mockText,
  }

  it('+++ should match snapshot', () => {
    const rendered = renderer.create(
      <ParagraphComponent
        id={mockId}
        text={mockText}
        language={mockLanguage}
        getTextResource={mockGetTextResource}
        textResourceBindings={mockTextResourceBindings}
      />,
    );
    expect(rendered).toMatchSnapshot();
  });
  it('+++ should have correct text', () => {
    const shallowParagraphComponent = shallow(
      <ParagraphComponent
        id={mockId}
        text={'some other text'}
        language={mockLanguage}
        getTextResource={mockGetTextResource}
        textResourceBindings={mockTextResourceBindings}
      />,
    );

    expect(shallowParagraphComponent.text()).toEqual('some other text');
  });

  it('+++ should render help text if help text is supplied', () => {
    const shallowParagraphComponent = shallow(
      <ParagraphComponent
        id={mockId}
        text={mockText}
        language={mockLanguage}
        getTextResource={mockGetTextResource}
        textResourceBindings={{ help: 'this is the help text'}}
      />,
    );

    expect(shallowParagraphComponent.find('HelpTextContainer')).toHaveLength(1);
  });
});
