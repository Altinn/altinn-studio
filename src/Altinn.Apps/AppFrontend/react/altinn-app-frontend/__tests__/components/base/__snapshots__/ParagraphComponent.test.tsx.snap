// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> components/base/ParagraphComponent.tsx --- Snapshot +++ should match snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <p
      className="MuiTypography-root makeStyles-spacing-1 makeStyles-typography-2 MuiTypography-body1"
      id="mock-id"
    >
      Here goes a paragraph
    </p>
  </div>
</div>
`;
